File: app/src/main/java/org/doomer/qnotez/viewmodel/NoteListViewModel.java
package org.doomer.qnotez.viewmodel;

import android.app.Application;
import android.arch.lifecycle.AndroidViewModel;
import android.arch.lifecycle.LiveData;
import android.util.Log;

import org.doomer.qnotez.db.AppDatabase;
import org.doomer.qnotez.db.NoteModel;
import org.doomer.qnotez.utils.NoteUtils;
import org.doomer.qnotez.utils.TextUtils;

import java.util.List;
import java.util.concurrent.ExecutionException;

public class NoteListViewModel extends AndroidViewModel {

    private  LiveData<List<NoteModel>> noteItems;
//    private MutableLiveData<List<NoteModel>> noteItems = new MutableLiveData<>();
    protected AppDatabase database;
    private boolean showInTrash = false;

    public NoteListViewModel(Application application) {
        super(application);

        database = AppDatabase.getDatabase(this.getApplication());
        noteItems = database.getNoteModel().getAllItems();

//        noteItems.setValue(database.getNoteModel().getAllItems().getValue());
    }

    public LiveData<List<NoteModel>> getNoteItems() {

        if (!showInTrash) {
            noteItems = database.getNoteModel().getAllItems();
        } else {
            noteItems = database.getNoteModel().getAllItemsInTrahs();
        }

        return noteItems;
    }

    public LiveData<List<NoteModel>> quickSearch(String text, boolean inTrash) {
        text = TextUtils.prepareToLikeQuery(text);

        String inTrashStr = inTrash ? "1" : "0";
        NoteUtils.SearchAsyncTask searchTask = new NoteUtils.SearchAsyncTask(database);

        try {
            noteItems = searchTask.execute(text, inTrashStr).get();
        } catch (InterruptedException e){
            Log.d("SSSSS", "INTERRUPT EXCEPTION");
        } catch (ExecutionException e) {
            Log.d("SSSSS", "EXECUTE EXCEPTION");
        }

        return noteItems;
    }

    public void deleteItem(NoteModel item) {
        new NoteUtils.NoteDeleteAsyncTask(database).execute(item);
    }


    public void setShowTrash(boolean trashVisible) {
        showInTrash = trashVisible;
    }

    public void moveToTrash(NoteModel item) {
        item.setInTrash(true);
        new NoteUtils.NoteUpdateAsyncTask(database).execute(item);
    }

    public void moveFromTrash(NoteModel item) {
        item.setInTrash(false);
        new NoteUtils.NoteUpdateAsyncTask(database).execute(item);
    }
}

File: app/src/main/kotlin/org/doomer/qnotez/viewmodel/NoteListViewModel.kt
package org.doomer.qnotez.viewmodel

import android.arch.lifecycle.LiveData
import android.arch.lifecycle.ViewModel
import android.util.Log

import org.doomer.qnotez.db.AppDatabase
import org.doomer.qnotez.db.NoteModel
import org.doomer.qnotez.utils.NoteUtils
import org.doomer.qnotez.utils.TextUtils
import java.util.concurrent.ExecutionException
import javax.inject.Inject

class NoteListViewModel @Inject constructor(private val db : AppDatabase) : ViewModel() {

    private var noteItems: LiveData<List<NoteModel>> //? = null
    private var showInTrash = false

    init {
        noteItems = db.noteModel().allItems
    }

    fun getNoteItems(): LiveData<List<NoteModel>> {

        if (!showInTrash) {
            noteItems = db.noteModel().allItems
        } else {
            noteItems = db.noteModel().allItemsInTrash
        }

        return noteItems
    }

    fun quickSearch(text: String, inTrash: Boolean): LiveData<List<NoteModel>> {
        var text = text
        text = TextUtils.prepareToLikeQuery(text)

        val inTrashStr = if (inTrash) "1" else "0"
        val searchTask = NoteUtils.SearchAsyncTask(db)

        try {
            noteItems = searchTask.execute(text, inTrashStr).get()
        } catch (e: InterruptedException) {
            Log.d("QNotez::QSearch", "INTERRUPT EXCEPTION")
        } catch (e: ExecutionException) {
            Log.d("QNotez::QSearch", "EXECUTE EXCEPTION")
        }

        return noteItems
    }

    fun addItem(item: NoteModel) {
        NoteUtils.NoteAddAsyncTask(db).execute(item)
    }

    fun deleteItem(item: NoteModel) {
        NoteUtils.NoteDeleteAsyncTask(db).execute(item)
    }


    fun cleanTrash() {
        NoteUtils.CleanTrashAsyncTask(db).execute("")
    }

    fun setShowTrash(trashVisible: Boolean) {
        showInTrash = trashVisible
    }

    fun moveToTrash(item: NoteModel) {
        item.isInTrash = true
        NoteUtils.NoteUpdateAsyncTask(db).execute(item)
    }

    fun moveFromTrash(item: NoteModel) {
        item.isInTrash = false
        NoteUtils.NoteUpdateAsyncTask(db).execute(item)
    }

    fun notesForBackup() : List<NoteModel> {
        val searchTask = NoteUtils.BackupAsyncTask(db)

        var backupItems : List<NoteModel> = mutableListOf()
        try {
             backupItems = searchTask.execute("").get()
        } catch (e: InterruptedException) {
            Log.d("QNotez::QBackup", "INTERRUPT EXCEPTION")
        } catch (e: ExecutionException) {
            Log.d("QNotez::QBackup", "EXECUTE EXCEPTION")
        }

        return backupItems
    }

    fun countInTrash() : Int {

        val countTask = NoteUtils.CountTrashAsyncTask(db)

        var count : Int = 0
        try {
            count = countTask.execute("").get()
        } catch (e: InterruptedException) {
            Log.d("QNotez::QBackup", "INTERRUPT EXCEPTION")
        } catch (e: ExecutionException) {
            Log.d("QNotez::QBackup", "EXECUTE EXCEPTION")
        }

        return count
    }
}


